### ISSUE TITLE
Account aliases.

### ISSUE THREAD

**Issue by ticki** [tootsuite/mastodon#3796]
The database should have a redirection field, which effectively redirects the trafic to the aliased account. For example, I have ticki@mastodon.social. Say I wanted to have ticki@redox-os.org redirecting to ticki@mastodon.social. It allows partial migration with relatively simple change.

Related to https://github.com/tootsuite/mastodon/issues/177

* * * *

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.
- [ ] This bug happens on a [tagged release](https://github.com/tootsuite/mastodon/releases) and not on `master` (If you're a user, don't worry about this).

**Comment by kerim** [comment]
Yes please! One way Mastodon was explained to me was that accounts on different instances are like email addresses, but in email I can set up aliases so that even if I send an email from a Gmail address, it appears to come from my work address. I see a lot of smaller instances popping up as kinds of "Facebook Groups" around different topics, but having to create a new identity for each group creates problems since I can't sync the follows (or followers) from each group, and I get separate notifications for conversations under each account. This could get tiresome very fast, but having the ability to create a local account with an alias would solve all of that - with all my social connections and replies going back to my primary account. Of course, for the purpose of moderation there should be a way to see if people are using an alias (just as one can do if one looks at email headers), but this needn't be obvious to the casual user.

**Comment by sethlsanders** [comment]
I agree that the alias feature would be a big factor in making Mastodon useful and attractive for me.

**Comment by glintinspace** [comment]
This would help keep people from harassing others by impersonating them.

**Comment by PubliqPhirm** [comment]
This would also be a good way to help implement #2668 (hosting several domains of instances from the same backend instance)

**Comment by duck57** [comment]
If you control all those domains, would CNAMES help or am I misunderstanding the issue?

**Comment by roipoussiere** [comment]
Instead I think it could be done using DNS entries.

As I understand well, I'm thinking about 3 entries:
- A `CNAME` entry, to redirect the domain;
- a `TXT` entry, to specify the username in the target instance, and a verification token.

These two values could be generated in the user configuration page (as GitLab does with custom-domain pages).

**Comment by remram44** [comment]
@roipoussiere This means that Mastodon needs access to the DNS zone, vastly complicating installation, and requiring a lot of code to be written to support the multitude of DNS providers, each with their own DNS updating API.

It would also mean that any account lookup needs to first to a DNS query before doing the current dance (webfinger, ActivityPub).

It would also vastly depart from the current standardized protocols (this is nowhere in the ActivityPub spec)

**Comment by roipoussiere** [comment]
> and requiring a lot of code to be written to support the multitude of DNS providers, each with their own DNS updating API.

Sorry for my lack of knowledge in this field, but why it's necessary to use DNS updating APIs?

**Comment by remram44** [comment]
When a user sets the alias on the Mastodon app, Mastodon would need to create the `TXT` entry you describe.

**Comment by roipoussiere** [comment]
Not necessary: the UI could just provide the DNS entries and the user manually adds them in the DNS provider interface.

**Comment by remram44** [comment]
The instance admin might be able to create DNS entry for his instance's domain, but that is certainly not the case for users.

**Comment by roipoussiere** [comment]
GitLab works like that, it displays DNS entries to the user, not to the GitLab instance admin.

**Comment by remram44** [comment]
... and the GitLab software uses some API to update the gitlab.io zone when users create accounts or projects.

**Comment by remram44** [comment]
GitLab doesn't need complicated TXT records, so they just [have a wildcard `*.gitlab.io` that points to their server](https://docs.gitlab.com/ee/administration/pages/#dns-configuration), and they never need to update it. This is very different from the scheme you described.

**Comment by galaxor** [comment]
I have another use-case for this feature, which could not be solved by DNS.

I have a server of my own, where I host my email.  So my email address is galaxor@myserver.example.com.  I also have a website up at myserver.example.com.  I know someone else who runs a mastodon server, theirserver.mastodon.example.com.  I would like to join their server so I don't have to run my own.  But if I do that, then I get a handle like galaxor@theirserver.mastodon.example.com.  Now, what if my friend stops running that instance?  What if I decide to run my own?  All my friends know me as galaxor@theirserver.mastodon.example.com, and I'm locked in.

Now, I could rectify this situation by hand-writing a webfinger file and hosting it at https://myserver.example.com/.well-known/webfinger?resource=galaxor@myserver.example.com

It could list all the appropriate aliases, profile-page, updates-from, salmon addresses, etc for theirserver.mastodon.example.com.  Problem solved, right??

Sort of:  I can tell people my webfinger address as galaxor@myserver.example.com (same as my email address -- appropriate).  However, if they actually subscribe to me, all my messages will be listed as coming from galaxor@theirserver.mastodon.example.com.  I need a way to inform theirserver.mastodon.example.com that my canonical address is galaxor@myserver.example.com.

In email, this is accomplished by simply changing the From: header.  Even gmail has a UI for that, so you can use gmail but have your address that everybody sees be galaxor@myserver.example.com.

Mastodon should have a thing like that!

**Comment by rkingett** [comment]
Yes! It would also save having to create a new name and password for each and every instance I want to join. I do find the [moving accounts feature](https://docs.joinmastodon.org/user/moving/) very helpful though.

**Comment by Daniel-Abrecht** [comment]
[I need this too](https://github.com/mastodon/mastodon/issues/3796#issuecomment-451274118)

I want to use mastodon, but I don't want to lock myself into a specific instance. I already made an account on one instance once, and later wasn't happy with some of it's restrictions, so I didn't announce it to anyone. I could make an account on another instance, but will I be happy there, and will it stay up? I can't tell in advance. I can't just tell anyone "I'm there", if I don't know if I will really be found there. I do have my own server & domain, but I do not want to host mastodon myself. I'm told some instances are picky and won't federate with instances with a single user, so that's not an option.

So I really do need some way to specify, use *that* instance, but this handle is my primary one.

**Comment by jeffkaufman** [comment]
@galaxor 's proposal sounds really good to me!  (reasoning: [post](https://www.jefftk.com/p/better-mastodon-aliases), [toot](https://mastodon.mit.edu/@jefftk/109347720341123283))

**Comment by erlend-sh** [comment]
This would make it a lot easier for people to properly dive into the local experience of various instances they’re curious about outside of their primary pick.

Feels similar to GitHub’s [multiple email support](https://docs.github.com/en/account-and-profile/setting-up-and-managing-your-personal-account-on-github/managing-email-preferences/adding-an-email-address-to-your-github-account).

![Uploading B2948F3F-C043-41C6-B4EA-AA93AF5EEF1C.jpeg…]()

**Comment by erlend-sh** [comment]
This would make it a lot easier for people to properly dive into the local experience of various instances they’re curious about outside of their primary pick.

Feels similar to GitHub’s [multiple email support](https://docs.github.com/en/account-and-profile/setting-up-and-managing-your-personal-account-on-github/managing-email-preferences/adding-an-email-address-to-your-github-account).

![B2948F3F-C043-41C6-B4EA-AA93AF5EEF1C](https://user-images.githubusercontent.com/583842/227365139-c91a7bcb-4458-4414-8b1a-2e86aa46ad99.jpeg)

**Comment by agharbeia** [comment]
> @roipoussiere This means that Mastodon needs access to the DNS zone, vastly complicating installation, and requiring a lot of code to be written to support the multitude of DNS providers, each with their own DNS updating API.
> 
> It would also mean that any account lookup needs to first do a DNS query before doing the current dance (webfinger, ActivityPub), adding latency.
> 
> It would also vastly depart from the current standardized protocols (this is nowhere in the ActivityPub spec).

Does Mastodon now do the initial DNS configuration necessary for running an instance on a given domain?

Many services that require some DNS configuration simply describe the needed records and their expected values, and leave it to the administrator to do the actual settings.

**Comment by agharbeia** [comment]
@remram44
Right. I hadn't understood that part. It's unpractical.
A DNS zone cannot have thousands of records.
